### PR TITLE
ENH Uses unique pointers

### PR DESCRIPTION
**Pull request by thomasjpfan**
`neigh_distances` and `neigh_indices` can be a unique pointer.

The [shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) constructor allows casting from unique to shared (item 13).

CC @jjerphan

### CODE REVIEW

File: sklearn/metrics/_pairwise_distances_reduction.pyx [hunk 1]
```diff
@@ -1213,8 +1213,8 @@ cdef class PairwiseDistancesRadiusNeighborhood(PairwiseDistancesReduction):
         # Shared pointers (defined via shared_ptr) are use for safer memory management.
         # Unique pointers (defined via unique_ptr) can't be used as datastructures
         # are shared across threads for parallel_on_X; see _parallel_on_X_init_chunk.
```
Comment (jjerphan) [inline sklearn/metrics/_pairwise_distances_reduction.pyx:1215]:
```suggestion
        # Unique pointers (defined via unique_ptr) are also used for the main datastructures
        # and are cast as shared pointers across threads for parallel_on_X; see _parallel_on_X_init_chunk.
```

**Comment by jjerphan** [comment]
Thank you for the two suggestions, @thomasjpfan.

**Comment by jjerphan** [comment]
In fact item 13 is a [move constructor](https://en.cppreference.com/w/cpp/language/move_constructor) which must take a r-value reference (see the `&&` for item 13) in order to be used.

We get a compile error because we aren't using r-value here, but `self.neigh_{distance,indices}` which are l-value.
I guess it makes sense not the be able to cast `unique_ptr` to `shared_ptr` because this would just break the uniqueness of the pointer. This was my prior understanding before this PR, and it seems that the problem we are meeting here likely illustrates it.

Is this right or am I missing something?

I will just revert for now to have the CI pass on the original PR, still I am interested in pursuing the discussions here, @thomasjpfan.

**Comment by thomasjpfan** [comment]
My mental model was incorrect about the interaction between shared + unique pointers. Yes, everything needs to be shared.

**Comment by jjerphan** [comment]
It's fine. That's was an occasion to learn for me as well. :)

**Pull request merged**